examples/deal.II/bp.h

```diff
@@ -0,0 +1,1165 @@
+            const CeedScalar ug1 = ug[i + Q * 1 + Q * 3 * 2];
+            const CeedScalar ug2 = ug[i + Q * 2 + Q * 3 * 2];
+            vg[i + Q * 0 + Q * 3 * 2] =
+              qdata[i + Q * 0] * ug0 + qdata[i + Q * 5] * ug1 + qdata[i + Q * 4] * ug2;
+            vg[i + Q * 1 + Q * 3 * 2] =
+              qdata[i + Q * 5] * ug0 + qdata[i + Q * 1] * ug1 + qdata[i + Q * 3] * ug2;
+            vg[i + Q * 2 + Q * 3 * 2] =
+              qdata[i + Q * 4] * ug0 + qdata[i + Q * 3] * ug1 + qdata[i + Q * 2] * ug2;
+          }
+        }
+        break;
+    }
+  return 0;
+}
+
+
+
+/**
+ * Base class of operators.
+ */
+template <typename Number>
+class OperatorBase
+{
+public:
+  /**
+   * deal.II vector type
+   */
+  using VectorType = LinearAlgebra::distributed::Vector<Number>;
+
+  /**
+   * Initialize vector.
+   */
+  virtual void
+  reinit() = 0;
+
+  /**
+   * Perform matrix-vecor product
+   */
+  virtual void
+  vmult(VectorType &dst, const VectorType &src) const = 0;
+
+  /**
+   * Initialize vector.
+   */
+  virtual void
+  initialize_dof_vector(VectorType &vec) const = 0;
+
+  /**
+   * Compute inverse of diagonal.
+   */
+  virtual void
+  compute_inverse_diagonal(VectorType &diagonal) const = 0;
+};
+
+
+/**
+ * Operator implementation using libCEED.
+ */
+template <int dim, typename Number>
+class OperatorCeed : public OperatorBase<Number>
+{
+public:
+  using VectorType = typename OperatorBase<Number>::VectorType;
+
+  /**
+   * Constructor.
+   */
+  OperatorCeed(const Mapping<dim>              &mapping,
+               const DoFHandler<dim>           &dof_handler,
+               const AffineConstraints<Number> &constraints,
+               const Quadrature<dim>           &quadrature,
+               const BPType                    &bp)
+    : mapping(mapping)
+    , dof_handler(dof_handler)
+    , constraints(constraints)
+    , quadrature(quadrature)
+    , bp(bp)
+  {
+    reinit();
+  }
+
+  /**
+   * Destructor.
+   */
+  ~OperatorCeed()
+  {
+    CeedOperatorDestroy(&op_apply);
+    CeedQFunctionDestroy(&qf_apply);
+    CeedQFunctionContextDestroy(&build_ctx);
+    CeedVectorDestroy(&q_data);
+    CeedElemRestrictionDestroy(&q_data_restriction);
+    CeedElemRestrictionDestroy(&sol_restriction);
+    CeedBasisDestroy(&sol_basis);
+    CeedDestroy(&ceed);
+  }
+
+  /**
+   * Initialized internal data structures, particularly, CEED.
+   */
+  void
+  reinit() override
+  {
+    const auto &tria = dof_handler.get_triangulation();
+    const auto &fe   = dof_handler.get_fe();
+
+    const auto n_components = fe.n_components();
+
+    if (bp == BPType::BP1 || bp == BPType::BP3 || bp == BPType::BP5)
+      {
+        AssertThrow(n_components == 1, ExcInternalError());
+      }
+    else
+      {
+        AssertThrow(n_components == dim, ExcInternalError());
+      }
+
+    // 1) create CEED instance -> "MatrixFree"
+    const char *ceed_spec = "/cpu/self/avx/blocked";
+    CeedInit(ceed_spec, &ceed);
```

Comment:
> How do we set the libCEED backend when we run the test?

This is hard coded for now. But this can be changed.

Reply:
This would go in the `Parameters` struct to make it runtime selectable if I'm understanding correctly? Selecting the libCEED backend resource at runtime is the big change needed to get this working in our CI pipeline.